### ISSUE TITLE
Winget manifest with github workflows for the new release.

### ISSUE THREAD

**Issue by bartoszek** [Slicer/Slicer#6651]
## Is your feature request related to a problem? Please describe.
Would be nice to have a `winget` manifest pushed to `winget-pkg` when publishing new release, auto update and command line unsupervised download/install on windows with `winget` is a hudge quality of life improvement.
Example implementation: https://github.com/sn4k3/UVtools/blob/.github/workflows/package-publish.yml#L16

**Comment by bartoszek** [comment]
Nope, I'm specifically thinking about 3dslicer, have some CT scan to mesh/print and no access to my Linux box. Sadly find out that on my Win10 laptop package for `3dslicer` is missing from `winget`, `choco` have it but still it would be nice to congruentalize the packaging to the defacto default one for the windows system.

**Comment by lassoan** [comment]
It could be nice if Slicer were available on package managers, it is just a matter of priorities and yet another few steps to the already convoluted release process.

Chocolatey does not seem to worth the effort, as [setting up of chocolatey itself](https://chocolatey.org/install) is much more complicated than installing Slicer. I know that people may already have it installed, but that is such a small number of users that it would probably not worth the trouble. If someone does it for us, then it's great, but I would not add that to the release checklist.

[It seems that Microsoft takes winget seriously enough](https://learn.microsoft.com/en-us/windows/package-manager/winget/) that it may worth attention of Slicer core developers. The [contribution process](https://github.com/microsoft/winget-pkgs) seems to be reasonable (just a github pull request) and it could be even somewhat automated.

It would be also nice to secure the `Slicer` or `3DSlicer` name, because if others claim it before us then it maybe harder to get it back.

@bartoszek I guess you are using winget, so you are somewhat more familiar with how it works, its conventions, etc. Would you be so kind and help us with submitting a pull request to https://github.com/microsoft/winget-pkgs for Slicer-5.0.3? It would be great if you could also write up a few sentences that we should add to our [release checklist](https://github.com/Slicer/Slicer/blob/main/.github/ISSUE_TEMPLATE/release.md) to keep the package up-to-date? Thank you!

**Comment by sn4k3** [comment]
> It would be also nice to secure the Slicer or 3DSlicer name, because if others claim it before us then it maybe harder to get it back.

winget requires a full name that is comapany.productname. And then multiple applications with same name can co-exist. If more than one app exists with same name a winget install <name> would return a list of possible full ids to use instead and use alias will not be possible. So reserving a name is not a thing here.

After create the manifest files with **wingetcreate**, the update PR is quite simple and often very easy that takes one line, eg: `wingetcreate.exe update PTRTECH.UVtools --urls "$msiUrl|x64" --version $version --token $wingetToken --submit`
it will fetch URL and auto submit a PR.

**Comment by jcfr** [comment]
### manifest

> create the manifest files with `wingetcreate` [...] it will fetch URL, calculate checksum and auto submit a PR

**Question:** Out of curiosity, is there way to specify the expected checksum ? Or (a easy way to) verify that the computed one match our record ?

---------

### automatic pull-request

> static url  [...] winget bot will detect checksum mismatch and auto PR with latest version

Since we have full control of https://download.slicer.org, we could support this.

It turns out we already have an endpoint for downloading the latest available package for a specific operating system and "stability":

```
https://download.slicer.org/download?os=<operating_system>&stability=<stability>
```

where

* operating_system` is `win`, `macosx` or `linux`
* stability is `release` or `nightly`

For example: `https://download.slicer.org/download?os=win&stability=release`

Notes:
* It that doesn't work, it would be a matter of improving the existing[^2] `/download` endpoint.
* We are planing to expand this to support `arch` parameter[^3]

[^2]: https://github.com/Slicer/slicer_download/blob/60bdadfd0ec34ad756a623616437804bf42269f9/slicer_download_server/__init__.py#L97-L98

[^3]: https://github.com/Slicer/slicer_download/issues/22

**Questions:**
1. Is the URL of the form `/download?os=<operating_system>&stability=<stability>` valid ?
2. If available, where is the documentation discussing with automatic pull request creation ?

---------

### package name

> in this case would be `Slicer.Slicer`.

This is indeed consistent with how the Slicer GitHub organization is laid out and the value of the macOS package bundle identifier.

Note there there are inconsistencies[^1] with how the settings file are organized.

> secure the `Slicer` or `3DSlicer` name

**Questions:** Should we submit both? Is there support for creating an "alias".

[^1]: https://discourse.slicer.org/t/inconsistency-between-organization-name-used-for-settings-files-and-for-macos-bundle-identifier/26547

----------

### GitHub workflow for package upload

Since the signing of the release package and the associated upload are down manually[^4], implementing a workflow like is done in `UVtools`[^5] will not work.

Instead, we should:
* rely on the automatic pull-request created by the winget bot (which would be checking the download URL as discussed above)

* create a GitHub workflow checking on a weekly basis if the current release is not yer available on winget and submit a pull request accordingly.

[^4]: https://github.com/Slicer/Slicer/blob/6834c51215cd3217e01da622c4d4687255a92bea/.github/ISSUE_TEMPLATE/release.md?plain=1#L51-L54
[^5]: https://github.com/sn4k3/UVtools/blob/master/.github/workflows/package-publish.yml

**Comment by sn4k3** [comment]
For reference see: [package main folder](https://github.com/microsoft/winget-pkgs/tree/master/manifests/p/PTRTECH/UVtools) and [package version/installer](https://github.com/microsoft/winget-pkgs/tree/master/manifests/p/PTRTECH/UVtools/3.9.0)

Your package main folder will contain all version which users can opt to install a older version when passing argument.
Each package version should include at least 3 yamal which define software metadata, locale and installer metadata. All of this is auto created/updated by wingetcreate.exe

> Question: Out of curiosity, is there way to specify the expected checksum ? Or (a easy way to) verify that the computed one match our record?

Checksum is calculated by winget (they download your installer and calculate) but you can also obtain same value, it's just the SHA256 of the installer.

> It turns out we already have an endpoint for downloading the latest available package for a specific operating system and "stability":

You can also support your "nightly" package, [example of such](https://github.com/microsoft/winget-pkgs/tree/master/manifests/a/ArduinoSA/IDE):

```
winget search arduino
Nome                                ID                   Versão      Correspondência Origem
--------------------------------------------------------------------------------------------
Arduino IDE                         9NBLGGH4RSD8         Unknown                     msstore
Remote Arduino Monitor              9NBLGGH4VBHQ         Unknown                     msstore
Windows Remote Arduino Experience   9NBLGGH2041M         Unknown                     msstore
Windows Virtual Shields for Arduino 9NBLGGGZ0MLD         Unknown                     msstore
Arduino IDE                         ArduinoSA.IDE.stable 2.0.0       Tag: arduino    winget
Arduino IDE RC                      ArduinoSA.IDE.rc     2.0.0-rc9.3 Tag: arduino    winget
Arduino IDE Beta                    ArduinoSA.IDE.beta   2.0.0.0                     winget
```

So your folder structure will be: `Slicer.Slicer.release` and `Slicer.Slicer.nightly` if you want to support both


> Is the URL of the form /download?os=<operating_system>&stability=<stability> valid ?

It is, you can use anykind of valid URL that lead to a direct file, redirect URLs are also valid. Javascript/delayed downloads aren't valid, eg: "Please wait 5 seconds before download or click here"

> If available, where is the documentation discussing with automatic pull request creation ?

Good question, I have [seen people using static link](https://github.com/microsoft/winget-pkgs/blob/master/manifests/t/the-sz/Compton/1.51/the-sz.Compton.installer.yaml) but I don't know how it's handled. In homebrew cask you can have a livecheck condition to fetch version from a page or whatever, on winget such does not exists. So my guess is when checksum does not match they fetch version from installer file and update with such?
My deep recommendation is to use direct link for each version and issue a PR. Why? Because winget does not store your installer, it come from your page and if you have a static link for the latest users can "never?" install a older version using winget. Also will prevent users from installing latest version util winget issue a PR due mismatch checksum, unless they pass the a option to ignore such check.

> package name
in this case would be Slicer.Slicer.
[...]
secure the Slicer or 3DSlicer name
[...]
Questions: Should we submit both? Is there support for creating an "alias".

See: https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.singleton.1.0.0.json
`PackageIdentifier` and `Moniker` both are plain string, so no. And for the sake of redudancy and user I'm glad that they dont support alias. Winget is all about "keep it simple".

> GitHub workflow for package upload
[...]
Instead, we should:

[...]

I recommend the GitHub workflow cron or a manuall trigger workflow. Many mantainers do the github workflow to support a bunch of 3th party apps which they don't own. Anyone can contribute and deliver the manifest updates.

I also have an powershell script in my PC that I use to trigger a PR using wingetcreate if my workflow fails by any reason

## Bonus

After place your software on winget, even if user install manually by your page download, winget will recognize the installation and users can simply do `winget upgrade Slicer.Slicer` when they want to update for the latest.
For example, see this software i have manually installed on my system (without winget):

```
winget upgrade
Nome                                           ID                                Versão           Disponível     Origem
-----------------------------------------------------------------------------------------------------------------------
010 Editor 13.0 (64-bit)                       SweetScape.010Editor              13.0             13.0.1         winget
```

So winget is very worth and trouble free.


Start here: https://learn.microsoft.com/en-us/windows/package-manager/package/manifest?tabs=minschema%2Cversion-example
